### PR TITLE
Change the default phase 2 PF reconstruction for endcap from simPF to pfTICL

### PR DESCRIPTION
**Pull request by hatakeyamak**
#### PR description:

This PR will change the default phase 2 PF reconstruction for endcap from simPF to pfTICL. This is based on the agreed plan between the PF group and HGCAL DPG that we will make this switch when #32291 is integrated (which happened a couple of days ago). The simPFProducer is still run for now, but not injected to the PF candidate list.

#### PR validation:

I ran this PR with the default setting (pfTICL) and also with `--customise RecoParticleFlow/Configuration/RecoParticleFlow_cff.replaceTICLwithSimPF`, and compared PF candidate outputs (actually packedCandidates) with those without this PR with the default (SimPF) and pfTICL injected. The test was done with 23234.0_TTbar 2026D49 without PU. I confirmed identical results when pfTICL or simPF is used for both without and with this PR.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport.

@bendavid @rovere @felicepantaleo @jnsandhya @zeratul87

### CODE REVIEW

**Comment by srimanob** [comment]
Please test

**Comment by slava77** [comment]
enable profiling

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Some remarks:
1) significantly fewer PFCandidates produced in the range 1.8<|eta|<3.0: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_3_X_2021-01-28-2300+0416e2/41030/validateJR/all_OldVSNew_TTbar14TeV2026D49wf23234p0/all_OldVSNew_TTbar14TeV2026D49wf23234p0c_recoPFCandidates_particleFlow__RECO_obj_eta.png, consequently downstream quantities also change. @hatakeyamak can you confirm this is expected from the change?
2) only 2026D60 and 2026D49 workflows are affected
4) miniAOD size in 23434.21 goes up by ~3%, driven by patIsolatedTracks_isolatedTracks__PAT: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0416e2/12600/profiling/23434.21/products_miniAOD_sizes_compare_23434.21.txt, not clear how this is consistent with 1)
3) AOD size doesn't change
5) code changes LGTM

**Comment by kpedro88** [comment]
Putting some plots here for posterity:

eta: 
image

pt (showing that the decrease is mostly at low pt):
image

**Comment by hatakeyamak** [comment]
@jpata @kpedro88 thank you.

I certainly see smaller number of PF candidates in my private test, I believe the changes are mostly coming from lowpt neutrals which are sensitive to reconstruction details. @kpedro88 can the above plots be made for charged candidates and neutral candidates separately?

> miniAOD size in 23434.21 goes up by ~3%, driven by patIsolatedTracks_isolatedTracks__PAT: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0416e2/12600/profiling/23434.21/products_miniAOD_sizes_compare_23434.21.txt, not clear how this is consistent with 1)

@jpata I think this is consistent with the fact that there a little reduced neutrals, which lead to somewhat increased number of isolated tracks.

I wonder @rovere has any additional comments/insight to offer.

**Comment by kpedro88** [comment]
Here are the available plots generated for the comparison (of course, one can also download the output files from the matrix tests and make any plots):

<details>
<summary>pt (charged hadron)</summary>

image
</details>

<details>
<summary>pt (electron)</summary>

image
</details>

<details>
<summary>pt (photon)</summary>

image
</details>

<details>
<summary>pt (neutral hadron)</summary>

image
</details>

<details>
<summary>pt (HF hadron)</summary>

image
</details>

The changes are indeed concentrated in the photon and neutral hadron categories.

**Comment by hatakeyamak** [comment]
@kpedro88 Thanks. Right. I should have just looked at automatically produced plots I guess, but thanks for putting the links above. Differences look like what I expected.

File: RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py [hunk 1]
```diff
@@ -84,11 +99,46 @@
...
+    ret = []
+    if getattr(process,'particleFlowBlock', None):
+        for i, pset in enumerate(process.particleFlowBlock.elementImporters):
+            if pset.importerName.value() == name:
+                ret.append(i)
+    return ret
+
+from Configuration.Eras.Modifier_phase2_hgcal_cff import phase2_hgcal
+def replaceTICLwithSimPF(process):
+    if getattr(process,'particleFlowTmp', None):
```
Comment (kpedro88) [inline RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py:124]: `hasattr(process,'particleFlowTmp')` is slightly simpler

**Comment by kpedro88** [comment]
@rovere does this PR mean that the `fromMultiCl` "parallel universe" for egamma reconstruction can be removed?

**Comment by hatakeyamak** [comment]
@kpedro88 Of course @rovere @SohamBhattacharya should know better, but if I understand correctly,  "fromMulticl" is used with ticl inputs now for egamma reco I think.
https://cmssdt.cern.ch/lxr/source/RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHGC_cfi.py

Here I was tracing "fromMulticl" sequence a few days ago.
https://docs.google.com/document/d/1fkgz7fPOatN2hkZ6X35rDnoMq0aJX1NtVmjKyQGuFBM/edit

**Comment by kpedro88** [comment]
@hatakeyamak yes: `FromMultiCl` was a "parallel universe" introduced for interim studies. If TICL becomes the default reco, it should no longer be needed: the "real" (not sim-cheated) HGCal objects will naturally be fed to the downstream producers. (Of course, there were some C++ algorithm changes needed when `FromMultiCl` was introduced, so the related settings might need to be propagated correctly to the default collections.)

Related question: should https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHGC_cfi.py be updated here?

**Comment by hatakeyamak** [comment]
I am not sure how ecalDrivenGsfElectronCoresFromMultiCl relates to electron candidates from pfticl, for example, and I wanted to get a big picture here. I will be happy to learn fro @SohamBhattacharya @rovere @felicepantaleo et al on this topic.

File: RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py (hunk 1 above)
Comment (hatakeyamak) [reply RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py:124]: If I don't hear any other suggestions, i will make this change soon.

**Comment by hatakeyamak** [comment]
There was a discussion of actually turning off SimPFProducer, but it will require a change in the validation modules, and I am still waiting feedback about it. I think such further change can be done in a later prerelease.

**Comment by SohamBhattacharya** [comment]
What I am confused about is, will I get the same electron candidates with pfTICL-based HGCal objects being fed downstream to the producers, as I get with using the TICL clusters in [1].
@kpedro88 Can you please point me to the location where the pfTICL-based HGCal objects will be entering the electron reconstruction chain?

[1] https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHGC_cfi.py#L51

**Comment by kpedro88** [comment]
@SohamBhattacharya see my previous comments on this PR: I think there is some more work that needs to be done to complete the integration of TICL in a consistent way. We should hear from @rovere et al. about their plans for this.

**Comment by SohamBhattacharya** [comment]
> @SohamBhattacharya see my previous comments on this PR: I think there is some more work that needs to be done to complete the integration of TICL in a consistent way. We should hear from @rovere et al. about their plans for this.

Okay, with your comment that this requires more C++ changes, I understand.
Otherwise it was not clear to me how this PR alone would add the HGCal-based electrons to the default collection.
Thanks.

**Comment by jpata** [comment]
@hatakeyamak could you post some of the PF DQM plots (jet response etc) in the PR description?

**Comment by hatakeyamak** [comment]
> @hatakeyamak could you post some of the PF DQM plots (jet response etc) in the PR description?

Not specifically. But, for ttbar 2026, I checked that the updated PR corresponds to the `injectTICLintoPF` with this PR:
https://github.com/cms-sw/cmssw/pull/32291
by comparing printoout of PF candidates.
So, http://hep.baylor.edu/hatake/PFval/val_11_3_0_pre2_PFOverlaps/ticl_simpf_test/
should corresponds to updates we are expecting with this PR for PF outputs.

**Comment by jpata** [comment]
Thanks for these plots. Looking through them a bit, I see some changes to jet resolution:
http://hep.baylor.edu/hatake/PFval/val_11_3_0_pre2_PFOverlaps/ticl_simpf_test/FlatQCD_noPU_JetMETJetValidation/slimmedJets/PtRecoOverGen_GenPt_E.pdf
image

and correspondingly for jet response: http://hep.baylor.edu/hatake/PFval/val_11_3_0_pre2_PFOverlaps/ticl_simpf_test/FlatQCD_noPU_ParticleFlow/JetResponse/slimmedJets/JEC/response_196_272.pdf
image

Posting this here for documentation primarily.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by hatakeyamak** [comment]
Probably this could be highlighted from QCD PU200, showing improvement of jet pt resolution at high pt, but room for improvement at low pt.

Screen Shot 2021-02-05 at 9 30 37 AM

File: RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py (hunk 1 above)
Comment (hatakeyamak) [reply RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py]: Done (including two other lines).

**Comment by jpata** [comment]
@cmsbuild please test

(it looks like the tests aborted due to timeout)

**Comment by hatakeyamak** [comment]
Thank you. Hopefully it will go through this time.

> @cmsbuild please test
> (it looks like the tests aborted due to timeout)

**Comment by hatakeyamak** [comment]
Profiling test seems to take a long time. @jpata @slava77 how long does it usually take? Is this exceptionally long?

**Comment by jpata** [comment]
profiling does take longer generally, but the tests still should succeed. in order to make sure we make the deadline, let me kill the profiling and restart without (we already verified profiling outputs earlier)

**Comment by jpata** [comment]
@cmsbuild please abort

**Comment by jpata** [comment]
test parameters:
  - enable_tests = none

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by smuzaffar** [comment]
Iprofiling job again timeed out after 6 hours. For now I have explicitly marked the profiling status to `pass` so that you can get the pr test results

**Comment by smuzaffar** [comment]
@jpata , https://github.com/cms-sw/cmssw/pull/32766#issuecomment-775785175 was the correct way to disable the profiling but there was a bug due to which bot was not overriding the https://github.com/cms-sw/cmssw/pull/32766#issuecomment-769840102 request. This has been fixed now

**Comment by jpata** [comment]
+reconstruction
- changes the default phase 2 PF reconstruction for endcap from simPF to pfTICL
- reco changes observed: less neutrals and photons, changes to jet resolution, these are all expected
- miniAOD size in 23434.21 goes up by ~3%
- code changes are OK

**Comment by srimanob** [comment]
+Upgrade

OK for the code. But not clear on validation plan. What is the validation plan? 
Should validation based on the same RAW, and compare between simPF and pfTICL with pre-release? High statistics?

**Comment by hatakeyamak** [comment]
> OK for the code. But not clear on validation plan. What is the validation plan?
> Should validation based on the same RAW, and compare between simPF and pfTICL with pre-release? High statistics?

If this one goes into 11_3_0_pre3, the comparison w.r.t. pre2 will become a good validation for this change.

On Monday's PPD coordination, we requested:
+ if possible, flat QCD for phase 2: increase statistics from usual 50k to 100k events (both noPIU and PU)
+ add VBF H invisible with 200k events
for 11_3_0_pre3 and I think PdmV is looking into it.

**Comment by silviodonato** [comment]
+1

**Comment by gartung** [comment]
> Profiling job again timeed out after 6 hours. For now I have explicitly marked the profiling status to `pass` so that you can get the pr test results

@smuzaffar I added a timeout of 4 hours to the igprof commands in the profiling scripts.

**Pull request merged**